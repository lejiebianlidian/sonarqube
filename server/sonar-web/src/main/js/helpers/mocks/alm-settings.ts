/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import {
  AlmKeys,
  AlmSettingsInstance,
  AzureBindingDefinition,
  BitbucketBindingDefinition,
  GithubBindingDefinition,
  GitlabBindingDefinition,
  ProjectAlmBindingResponse,
  ProjectBitbucketBindingResponse
} from '../../types/alm-settings';

export function mockAlmSettingsInstance(
  overrides: Partial<AlmSettingsInstance> = {}
): AlmSettingsInstance {
  return {
    alm: AlmKeys.GitHub,
    key: 'key',
    ...overrides
  };
}

export function mockAzureBindingDefinition(
  overrides: Partial<AzureBindingDefinition> = {}
): AzureBindingDefinition {
  return {
    key: 'key',
    personalAccessToken: 'asdf1234',
    ...overrides
  };
}

export function mockBitbucketBindingDefinition(
  overrides: Partial<BitbucketBindingDefinition> = {}
): BitbucketBindingDefinition {
  return {
    key: 'key',
    personalAccessToken: 'asdf1234',
    url: 'http://bbs.enterprise.com',
    ...overrides
  };
}

export function mockGithubBindingDefinition(
  overrides: Partial<GithubBindingDefinition> = {}
): GithubBindingDefinition {
  return {
    key: 'key',
    url: 'http://github.enterprise.com',
    appId: '123456',
    clientId: 'client1',
    clientSecret: '**clientsecret**',
    privateKey: 'asdf1234',
    ...overrides
  };
}

export function mockGitlabBindingDefinition(
  overrides: Partial<GitlabBindingDefinition> = {}
): GitlabBindingDefinition {
  return {
    key: 'foo',
    personalAccessToken: 'foobar',
    ...overrides
  };
}

export function mockProjectAlmBindingResponse(
  overrides: Partial<ProjectAlmBindingResponse> = {}
): ProjectAlmBindingResponse {
  return {
    alm: AlmKeys.GitHub,
    key: 'foo',
    ...overrides
  };
}

export function mockProjectBitbucketBindingGet(
  overrides: Partial<ProjectBitbucketBindingResponse> = {}
): ProjectBitbucketBindingResponse {
  return {
    alm: AlmKeys.Bitbucket,
    key: 'foo',
    repository: 'PROJECT_KEY',
    slug: 'repo-slug',
    ...overrides
  };
}
