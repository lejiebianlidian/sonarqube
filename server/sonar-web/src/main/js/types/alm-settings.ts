/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export const enum AlmKeys {
  Azure = 'azure',
  Bitbucket = 'bitbucket',
  GitHub = 'github',
  GitLab = 'gitlab'
}

export interface AlmBindingDefinition {
  key: string;
}

export interface AzureBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
}

export interface BitbucketBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
  url: string;
}

export interface GithubBindingDefinition extends AlmBindingDefinition {
  appId: string;
  clientId: string;
  clientSecret: string;
  privateKey: string;
  url: string;
}

export interface GitlabBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
  url?: string;
}

export interface ProjectAlmBindingResponse {
  alm: AlmKeys;
  key: string;
  repository?: string;
  slug?: string;
  summaryCommentEnabled?: boolean;
}

export interface ProjectBitbucketBindingResponse extends ProjectAlmBindingResponse {
  alm: AlmKeys.Bitbucket;
  repository: string;
  slug: string;
}

export interface ProjectAlmBindingParams {
  almSetting: string;
  project: string;
}

export interface AzureProjectAlmBindingParams extends ProjectAlmBindingParams {}

export interface BitbucketProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository: string;
  slug: string;
}

export interface GithubProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository: string;
  summaryCommentEnabled: boolean;
}

export interface GitlabProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository?: string;
}

export interface AlmSettingsInstance {
  alm: AlmKeys;
  key: string;
  url?: string;
}

export interface AlmSettingsBindingDefinitions {
  [AlmKeys.Azure]: AzureBindingDefinition[];
  [AlmKeys.Bitbucket]: BitbucketBindingDefinition[];
  [AlmKeys.GitHub]: GithubBindingDefinition[];
  [AlmKeys.GitLab]: GitlabBindingDefinition[];
}
