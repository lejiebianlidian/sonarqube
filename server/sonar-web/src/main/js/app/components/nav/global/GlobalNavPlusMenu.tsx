/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { Link } from 'react-router';
import { ButtonLink } from 'sonar-ui-common/components/controls/buttons';
import ChevronsIcon from 'sonar-ui-common/components/icons/ChevronsIcon';
import QualifierIcon from 'sonar-ui-common/components/icons/QualifierIcon';
import { translate } from 'sonar-ui-common/helpers/l10n';
import { getBaseUrl } from 'sonar-ui-common/helpers/urls';
import { ComponentQualifier } from '../../../../types/component';

export interface GlobalNavPlusMenuProps {
  canCreateApplication: boolean;
  canCreatePortfolio: boolean;
  canCreateProject: boolean;
  compatibleAlms: Array<string>;
  onComponentCreationClick: (componentQualifier: ComponentQualifier) => void;
}

function renderCreateProjectOptions(compatibleAlms: Array<string>) {
  return [...compatibleAlms, 'manual'].map(alm => (
    <li key={alm}>
      <Link
        className="display-flex-center"
        to={{ pathname: '/projects/create', query: { mode: alm } }}>
        {alm === 'manual' ? (
          <ChevronsIcon className="spacer-right" />
        ) : (
          <img
            alt={alm}
            className="spacer-right"
            width={16}
            src={`${getBaseUrl()}/images/alm/${alm}.svg`}
          />
        )}
        {translate('my_account.add_project', alm)}
      </Link>
    </li>
  ));
}

function renderCreateComponent(
  componentQualifier: ComponentQualifier,
  onClick: (qualifier: ComponentQualifier) => void
) {
  return (
    <li>
      <ButtonLink
        className="display-flex-justify-start padded-left"
        onClick={() => onClick(componentQualifier)}>
        <QualifierIcon className="spacer-right" qualifier={componentQualifier} />
        {translate('my_account.create_new', componentQualifier)}
      </ButtonLink>
    </li>
  );
}

export default function GlobalNavPlusMenu(props: GlobalNavPlusMenuProps) {
  const { canCreateApplication, canCreatePortfolio, canCreateProject, compatibleAlms } = props;

  return (
    <ul className="menu">
      {canCreateProject && (
        <>
          <li className="menu-header">
            <strong>{translate('my_account.add_project')}</strong>
          </li>
          {renderCreateProjectOptions(compatibleAlms)}
        </>
      )}
      {(canCreateApplication || canCreatePortfolio) && (
        <>
          {canCreateProject && <li className="divider" />}
          {canCreatePortfolio &&
            renderCreateComponent(ComponentQualifier.Portfolio, props.onComponentCreationClick)}
          {canCreateApplication &&
            renderCreateComponent(ComponentQualifier.Application, props.onComponentCreationClick)}
        </>
      )}
    </ul>
  );
}
