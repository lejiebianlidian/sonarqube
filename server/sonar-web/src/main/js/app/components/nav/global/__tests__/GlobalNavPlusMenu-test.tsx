/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { ButtonLink } from 'sonar-ui-common/components/controls/buttons';
import { AlmKeys } from '../../../../../types/alm-settings';
import { ComponentQualifier } from '../../../../../types/component';
import GlobalNavPlusMenu, { GlobalNavPlusMenuProps } from '../GlobalNavPlusMenu';

it('should render correctly', () => {
  expect(shallowRender({ canCreateApplication: true })).toMatchSnapshot('app only');
  expect(shallowRender({ canCreatePortfolio: true })).toMatchSnapshot('portfolio only');
  expect(shallowRender({ canCreateProject: true })).toMatchSnapshot('project only');
  expect(
    shallowRender({ canCreateProject: true, compatibleAlms: [AlmKeys.Bitbucket] })
  ).toMatchSnapshot('imports');
  expect(
    shallowRender({
      canCreateApplication: true,
      canCreatePortfolio: true,
      canCreateProject: true,
      compatibleAlms: [AlmKeys.Bitbucket]
    })
  ).toMatchSnapshot('all');
});

it('should trigger onClick', () => {
  const onComponentCreationClick = jest.fn();
  const wrapper = shallowRender({
    canCreateApplication: true,
    canCreatePortfolio: true,
    onComponentCreationClick
  });

  // Portfolio
  const portfolioButton = wrapper.find(ButtonLink).at(0);
  portfolioButton.simulate('click');
  expect(onComponentCreationClick).toBeCalledWith(ComponentQualifier.Portfolio);

  onComponentCreationClick.mockClear();

  // App
  const appButton = wrapper.find(ButtonLink).at(1);
  appButton.simulate('click');
  expect(onComponentCreationClick).toBeCalledWith(ComponentQualifier.Application);
});

function shallowRender(overrides: Partial<GlobalNavPlusMenuProps> = {}) {
  return shallow(
    <GlobalNavPlusMenu
      canCreateApplication={false}
      canCreatePortfolio={false}
      canCreateProject={false}
      compatibleAlms={[]}
      onComponentCreationClick={jest.fn()}
      {...overrides}
    />
  );
}
