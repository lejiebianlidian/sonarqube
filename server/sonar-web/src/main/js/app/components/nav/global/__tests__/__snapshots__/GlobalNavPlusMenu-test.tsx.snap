// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: all 1`] = `
<ul
  className="menu"
>
  <li
    className="menu-header"
  >
    <strong>
      my_account.add_project
    </strong>
  </li>
  <li
    key="bitbucket"
  >
    <Link
      className="display-flex-center"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/projects/create",
          "query": Object {
            "mode": "bitbucket",
          },
        }
      }
    >
      <img
        alt="bitbucket"
        className="spacer-right"
        src="/images/alm/bitbucket.svg"
        width={16}
      />
      my_account.add_project.bitbucket
    </Link>
  </li>
  <li
    key="manual"
  >
    <Link
      className="display-flex-center"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/projects/create",
          "query": Object {
            "mode": "manual",
          },
        }
      }
    >
      <ChevronsIcon
        className="spacer-right"
      />
      my_account.add_project.manual
    </Link>
  </li>
  <li
    className="divider"
  />
  <li>
    <ButtonLink
      className="display-flex-justify-start padded-left"
      onClick={[Function]}
    >
      <QualifierIcon
        className="spacer-right"
        qualifier="VW"
      />
      my_account.create_new.VW
    </ButtonLink>
  </li>
  <li>
    <ButtonLink
      className="display-flex-justify-start padded-left"
      onClick={[Function]}
    >
      <QualifierIcon
        className="spacer-right"
        qualifier="APP"
      />
      my_account.create_new.APP
    </ButtonLink>
  </li>
</ul>
`;

exports[`should render correctly: app only 1`] = `
<ul
  className="menu"
>
  <li>
    <ButtonLink
      className="display-flex-justify-start padded-left"
      onClick={[Function]}
    >
      <QualifierIcon
        className="spacer-right"
        qualifier="APP"
      />
      my_account.create_new.APP
    </ButtonLink>
  </li>
</ul>
`;

exports[`should render correctly: imports 1`] = `
<ul
  className="menu"
>
  <li
    className="menu-header"
  >
    <strong>
      my_account.add_project
    </strong>
  </li>
  <li
    key="bitbucket"
  >
    <Link
      className="display-flex-center"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/projects/create",
          "query": Object {
            "mode": "bitbucket",
          },
        }
      }
    >
      <img
        alt="bitbucket"
        className="spacer-right"
        src="/images/alm/bitbucket.svg"
        width={16}
      />
      my_account.add_project.bitbucket
    </Link>
  </li>
  <li
    key="manual"
  >
    <Link
      className="display-flex-center"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/projects/create",
          "query": Object {
            "mode": "manual",
          },
        }
      }
    >
      <ChevronsIcon
        className="spacer-right"
      />
      my_account.add_project.manual
    </Link>
  </li>
</ul>
`;

exports[`should render correctly: portfolio only 1`] = `
<ul
  className="menu"
>
  <li>
    <ButtonLink
      className="display-flex-justify-start padded-left"
      onClick={[Function]}
    >
      <QualifierIcon
        className="spacer-right"
        qualifier="VW"
      />
      my_account.create_new.VW
    </ButtonLink>
  </li>
</ul>
`;

exports[`should render correctly: project only 1`] = `
<ul
  className="menu"
>
  <li
    className="menu-header"
  >
    <strong>
      my_account.add_project
    </strong>
  </li>
  <li
    key="manual"
  >
    <Link
      className="display-flex-center"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/projects/create",
          "query": Object {
            "mode": "manual",
          },
        }
      }
    >
      <ChevronsIcon
        className="spacer-right"
      />
      my_account.add_project.manual
    </Link>
  </li>
</ul>
`;
